thewalrus/internal_modes/prepare_cov.py

```diff
@@ -0,0 +1,304 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Set of functions for forming a covariance matrix over multiple modes, based on overlaps of the squeezer schmidt modes
+"""
+
+from itertools import chain
+
+import numpy as np
+
+from ..decompositions import takagi
+from ..quantum import fidelity
+from ..symplectic import (
+    interferometer,
+    passive_transformation,
+    reduced_state,
+    squeezing,
+)
+
+
+def indexconversion(j, k, njs):
+    r"""
+    Converts the double index (j, k) into a single index = k + \sum_{l=0}^{j-1} njs[l],
+    with njs[0] defined as 0, njs[1] as the number of Schmidt modes in the first spatial mode, etc.
+
+    Args:
+        j (int): spatial mode index
+        k (int): Schmidt mode index
+        njs (array): number of Schmidt modes in each
+
+    Returns:
+        (int): converted index
+    """
+    return k + np.sum(njs[:j])
+
+
+def swap_matrix(M, R):
+    r"""
+    Computes the matrix that swaps the ordering of modes from grouped by orthonormal modes to grouped by spatial modes.
+    The inverse of the swap matrix is its transpose.
+
+    Args:
+        M (int): number of spatial modes
+        R (int): number of orthonormal modes in each spatial mode
+
+    Returns:
+        (array): M*R x M*R swap matrix
+    """
+    P = np.zeros((M * R, M * R), dtype=int)
+    in_modes = range(M * R)
+    out_modes = list(chain.from_iterable(range(i, M * R, M) for i in range(M)))
+    P[out_modes, in_modes] = 1
+    return P
+
+
+def O_matrix(F):
+    r"""The overlap matrix of all of the Schmidt modes in each spatial mode.
+
+    Args:
+        F (list[list/array[array]): List of arrays of the temporal modes of the Schmidt modes in each spatial mode, must be normalized and ordered by spatial mode.
+                                    Sufficient to put the output temporal modes from yellowsubmaring in a list ordered by spatial mode.
+    Returns:
+        (array): the overlap matrix
+    """
+    Fs = [f for fj in F for f in fj]
+    R = len(Fs)
+    O = np.eye(R, dtype=np.complex128)
+    for j in range(R):
+        for k in range(R):
+            O[j, k] = np.inner(Fs[j].conj() / np.linalg.norm(Fs[j]), Fs[k] / np.linalg.norm(Fs[k]))
+    return O
+
+
+def vacuum_state(M, hbar=2):
+    r"""
+    Returns the displacement vector and covariance matrix for an M mode vacuum state.
+
+    Args:
+        M (int): number of modes (spatial, Schmidt, other etc.)
+        hbar (float): (default 2) the value of :math:`\hbar` in the commutation relation :math:`[\x,\p]=i\hbar`
+
+    Returns:
+        tuple[array]: the displacement vector and covariance matrix of an M mode vacuum state
+    """
+
+    return np.zeros(2 * M), np.eye(2 * M) * hbar / 2
+
+
+# pylint: disable=too-many-branches
+def orthonormal_basis(rjs, O=None, F=None, thr=1e-3):
```

Comment:
If both O and F are given (ie not `None`) it looks like F is used and O is ignored. Should we add a warning that O is ignored or a check that they are compatible (ie O=O_matrix(F)) if both are given?

Reply:
Sure we can add that